File: Merge Sort.java
import java.util.*;
public class HelloWorld{

     public static void main(String []args){
        int[] l1 = new int[] {5,3,1,10,15,3,4, 5, 11};
        mergeSort(l1, l1.length);
        for(int i =0; i< l1.length;++i){
            System.out.print(l1[i]+ " ");
        }
     }
     
     public static void mergeSort(int[] arr, int length) {
        if(length < 2) {
            return;
        }
        int mid = length / 2;
        int[] left = new int[mid];
        int[] right = new int[length - mid];
        
        int k = 0;
        for(int i=0; i<length; i++) {
            if(i < mid) {
                left[i] = arr[i];
            } else {
                right[k++] = arr[i];
            }
        }
        
        mergeSort(left, mid);
        mergeSort(right, length - mid);
        merge(left, right, arr);
     }
     
     public static void merge(int[] l1, int[] l2, int[] arr) {
        int p1 = 0;
        int p2 = 0;
        int count = 0;
        while(p1 < l1.length && p2 < l2.length) {
            if(l1[p1] < l2[p2]) {
                arr[count] = l1[p1];
                p1++;
            } else {
                arr[count] = l2[p2];
                p2++;
            }
            count++;
        }
        while (p2 < l2.length) {
            arr[count++] = l2[p2++];
        }
        while (p1 < l1.length) {
            arr[count++] = l1[p1++];
        }
     }
}

File: Merge Sort.py
arr = [38, 27, 43, 3, 9, 82, 10]
print(arr)

def mergesort(arr):
    if len(arr) > 1:
        mid = len(arr)//2
        left = arr[:mid]
        right = arr[mid:]
        
        mergesort(left)
        mergesort(right)
        
        # Once the len(arr) == 1, the mergsesorts will stop and merging starts
        i, j, k = 0, 0, 0
        while i < len(left) and j < len(right):
            if left[i] <= right[j]:
                arr[k] = left[i]
                i += 1
            else:
                arr[k] = right[j]
                j += 1
            k += 1
        while i < len(left):
            arr[k] = left[i]
            i += 1
            k += 1
        while j < len(right):
            arr[k] = right[j]
            j += 1
            k += 1

mergesort(arr)
print(arr)
